### ISSUE TITLE
N5Store support of cloud buckets

### ISSUE THREAD

**Issue by chrisroat** [zarr-developers/zarr-python#540]
The N5Store works great for storing local files.  What are the technical challenges in moving to support something like GCS for storage?   Other pieces of the zarr library seamlessly work if I move to gs:// prefixes (with the right packages installed), but N5Store ends up writing to the local filesystem when using that prefix.

Is it a matter of going through n5.py and storage.py and changing uses of the os library to uses of fsspec?  Or are there operations that fundamentally won't work on cloud objects?

**Comment by jakirkham** [comment]
Funny this came up in conversation today. Also relates to issue ( https://github.com/zarr-developers/zarr-python/issues/395 ) and issue ( https://github.com/zarr-developers/zarr-python/issues/410 ).

cc @ryan-williams @joshmoore

**Comment by d-v-b** [comment]
I'm also affected by this issue -- I have n5 data on s3 (because that's what neuroglancer wants) but I'd also like to use zarr locally to access these data via s3fs. Concretely, what needs to be done to get this working?

**Comment by jakirkham** [comment]
This is probably not a difficult thing to solve if you are interested. As implemented [`N5Store`]( https://github.com/zarr-developers/zarr-python/blob/c959d11821da3b3e2b46be9dbac53e358d2a19a6/zarr/n5.py#L30 ) inherits from `NestedDirectoryStore`. However there is no reason it needs to do that. Instead we might consider giving `N5Store` a constructor that takes a `Store` as an argument. This would allow it to operate on other stores we might provide it. Internally we would replace `super` calls to using the `store` we gave it.

A similar thing may be needed for `NestedDirectoryStore`.

If we decide that we would rather not modify `N5Store` or `NestedDirectoryStore`, we could consider new classes `N5Mapping` and `NestedDirectoryMapping` to provide the same behavior as above, but in new classes. We could then simplify `N5Store` and `NestedDirectoryStore` to use these new `Mapping` classes.

Thoughts? 🙂

**Comment by d-v-b** [comment]
as a non-maintainer, either one looks great to me :) I will start with the first strategy and see how far I get.

**Comment by jakirkham** [comment]
Sounds good. Please let us know if there are other ways we can help 🙂

**Comment by chrisroat** [comment]
I like the first strategy, as well.  Meta-question: what sorts of Store's do you expect?  

The reason I ask is that the various implementations of fsspec could be used in a single FilesystemStore, which detects which fssspec implementation to use.  This avoids having a S3Store, GcsStore, LocalFsStore, etc.  Is this a worthwhile approach?

**Comment by mzjp2** [comment]
> I like the first strategy, as well. Meta-question: what sorts of Store's do you expect?
> 
> The reason I ask is that the various implementations of fsspec could be used in a single FilesystemStore, which detects which fssspec implementation to use. This avoids having a S3Store, GcsStore, LocalFsStore, etc. Is this a worthwhile approach?

Indeed, see @rabernat PR: https://github.com/zarr-developers/zarr-python/pull/373#issuecomment-592722584 -- I was going to try a quick implementation this weekend, happy to collaborate though!

**Comment by martindurant** [comment]
Let me know what you come up with @mzjp2 . It's also on my radar, but I don't know when.

**Comment by perlman** [comment]
Has anyone made progress on n5+http? I like @jakirkham's suggested approach, but figure my time would be better spent testing/tweaking someone else's code if it's hiding in a branch somewhere.

**Comment by martindurant** [comment]
It is on my list, but also happy to review rather than write!

**Comment by d-v-b** [comment]
it's on my list too :) but i haven't gotten to it yet

**Comment by chrisroat** [comment]
With #546 in the works, let me get my head around what we want here.  @alimanfoo has the good idea of re-writing N5Store to take a FSStore (or another store/mapping) and be simply a transformation.  It looks like this would mean simply replacing super calls with calls to the passed in mapping?

If I understand correctly, N5Store would then be a MutableMapping that takes another MutableMapping?

The code would look like:
```
fs_store = zarr.FSStore('gs://bucket/foo')
n5_store = zarr.N5Store(fs_store)  # Or rename this to zarr.N5Transformer
arr.to_zarr(n5_store)
```

**Comment by martindurant** [comment]
Sure, that would work fine.

However, it would also be really nice to have top level
`zarr.open_*(generic_url, storage_options=, is_n5, is_consolidated, is_nested)`
functions rather than have users passing around store objects. Maybe I'm asking too much, perhaps n5 is too different. 
On the other hand, this could be implemented via dispatch.

**Comment by shoyer** [comment]
> I have n5 data on s3 (because that's what neuroglancer wants)

Side note: Neuroglancer now supports directly zarr files, so that might be a good option for you!

**Comment by chrisroat** [comment]
Thanks for the heads up.  I'd like to migrate to that, as it'd be a bit more efficient than my current setup!  

I have a cloudvolume+igneous deployment going now -- the existing igneous code+config sets up downscaling pyramids out-of-the-box, which is super helpful and I'd have to replicate (maybe not hard... just time).

**Comment by joshmoore** [comment]
@martindurant : sorry, I missed the comment in #546 -- this isn't fully implemented yet, correct? But rather "NestedDirectoryStore support of cloud buckets", right?

**Comment by martindurant** [comment]
Correct, FSStore allows nesting by specifying the key separator. I don't know what more N5 needs, but presumably it could use of of these stores.

**Comment by joshmoore** [comment]
Ok. Re-opening this issue then.

**Comment by d-v-b** [comment]
I got FSStore-backed n5 writing working in my own repo: https://github.com/janelia-cosem/fibsem-tools/blob/master/src/fibsem_tools/io/storage.py

I created a class `N5FSStore` that inherits from `FSStore`, much like how `N5Store` inherits from `NestedDirectoryStore`
To get this to work, I had to modiify `FSStore` in two ways: 

* instead of defining the zarr meta keys (.zarray, .zgroup, .zattrs) as properties of the `FSStore` class, `FSStore` takes `meta_keys` as a keyword argument in its constructor. This allows subclasses to provide their own meta keys (e.g., `attributes.json` for n5).

* `FSStore.getitems` takes keys as input, transforms them via `_normalize_keys()`, and returns a dict of `{transformed_keys, values}`. This didn't work with n5 because the core zarr library asks for "." separated keys, which the n5 adapter code converts to "/"-separated keys in reverse order (e.g., ("0.1.2" -> "2/1/0"). If `getitems` returns a dict with the transformed keys, then the core zarr code doesn't recognize them and all the values are discarded. So I modified `getitems` to return a dict with the original keys instead. 

Beyond these two changes, I could recycle a lot of the existing key transformation logic from `N5Store` in `N5FSStore`. 

I need this for my own work so I haven't put together a PR yet. Would the changes I made to `FSStore` be acceptable? And would we want to have 2 N5 implementations (one derived from `NestedDirectoryStore`, and another derived from `FSStore`?

**Comment by joshmoore** [comment]
> * `FSStore` takes `meta_keys` as a keyword argument in its constructor

:+1:

> * So I modified `getitems` to return a dict with the original keys instead.

I don't know the impact of this offhand, but if tests are passing...

> Would the changes I made to `FSStore` be acceptable?

From my POV, having them would be great!

> would we want to have 2 N5 implementations

Since NestedDirectoryStore can't deal with remote access, I think we *minimally* need an FSStore version which can handle N5. (I had _some_ hope that we wouldn't need to subclass and instead could use composition, but I haven't looked into it further.)

I think the question is whether or not we maintain NDS or if it gets deprecated.

**Comment by martindurant** [comment]
I originally wrote FSStore to optionally contain N5 also, but it proved too unwieldy. I don't think there's a problem to have multiple store implementations. I haven't had a look at the changes yet, but passing the tests is a good indication.

**Comment by martindurant** [comment]
> I noticed that some of the tests for fsstore are incomplete

Please flesh out as you see fit; if this justifies the changes in FSStore, all the better.

**Comment by perlman** [comment]
I found myself trying to open a remote N5 this morning. I had completely forgotten about this issue. 🤦

**Comment by d-v-b** [comment]
@perlman see https://github.com/janelia-cosem/fibsem-tools/blob/master/src/fibsem_tools/io/storage.py#L223

Ultimately I want to take this code out of my own repo and get it here...

**Comment by martindurant** [comment]
And to be clear: code that works now is a great thing to have!

**Comment by perlman** [comment]
Thanks @d-v-b, I'll give it a whirl. I might have a few cycles to work on moving this over (...I would certainly use it).

> This has been mentioned elsewhere, but it may be better instead of subclassing, to have a layered approach, i.e., a N5 storage class that takes another storage class as an arguments.

@martindurant: Is there example to follow, style-wise, for doing this within python-zarr?

**Comment by joshmoore** [comment]
Not as far as I know. cc: @jakirkham who mentioned something similar to me as well in case he has an example.

**Comment by joshmoore** [comment]
@d-v-b: would you like to do the honors?

**Comment by d-v-b** [comment]
this should be addressed by #793

**Comment by joshmoore** [comment]
:100:
